clippy_lints/src/redundant_clone.rs

```diff
@@ -70,317 +69,729 @@ impl<'tcx> LateLintPass<'tcx> for RedundantClone {
+        for stmt in body.basic_blocks.iter().flat_map(|data| &*data.statements) {
+            if let StatementKind::Assign(a) = &stmt.kind
+                && let Rvalue::AddressOf(_, place) = a.1
+                && !place.is_indirect()
+            {
+                raw_ptr_taken.insert(place.local);
+            }
+        }
 
-        let mut possible_borrower = PossibleBorrowerMap::new(cx, mir);
+        let mut ref_target_results = RefTargetAnalysis::new(cx.tcx, cx.param_env, body, &raw_ptr_taken)
+            .into_engine(cx.tcx, body)
+            .iterate_to_fixpoint()
+            .into_results_cursor(body);
+        let clone_sources = clone_calls
+            .iter()
+            .filter_map(|block| {
+                let data = &body.basic_blocks[block];
+                ref_target_results.seek_before_primary_effect(Location {
+                    block,
+                    statement_index: data.statements.len(),
+                });
+                if let TerminatorKind::Call { args, .. } = &data.terminator().kind
+                    && let [Operand::Move(arg), ..] = &**args
+                    && let Some(Some(targets)) = ref_target_results.get().targets.get(&arg.local)
+                {
+                    Some((block, (arg.local, targets.clone())))
+                } else {
+                    None
+                }
+            })
+            .collect::<FxHashMap<_, _>>();
+        if clone_sources.is_empty() {
+            return;
+        }
 
-        for (bb, bbdata) in mir.basic_blocks.iter_enumerated() {
-            let terminator = bbdata.terminator();
+        let mut results = CloneAnalysis::new(cx, body, &raw_ptr_taken, &clone_sources)
+            .into_engine(cx.tcx, body)
+            .iterate_to_fixpoint();
+
+        // Check if a merged values are borrowed at the start of each block.
+        // Workaround for `join` not having access to the analysis.
+        let mut required_clones = take(&mut results.analysis.required_clones);
+        let mut borrowers = PossibleBorrowerMap::new(cx, body);
+        for block in (0..body.basic_blocks.len()).map(BasicBlock::from_usize) {
+            let state = results.entry_set_for_block(block);
+            let loc = block.start_location();
+            for (&local, _) in state.values.iter().filter(|(_, v)| v.last_modified.is_none()) {
+                for &(l1, l2, block) in &state.links {
+                    let local = if l1 == local {
+                        l2
+                    } else if l2 == local {
+                        l1
+                    } else {
+                        continue;
+                    };
 
-            if terminator.source_info.span.from_expansion() {
-                continue;
+                    if !borrowers.bounded_borrowers(&[], &[], local, loc) {
+                        required_clones.insert(block);
+                    }
+                }
             }
+        }
```

Comment:
The `required_clones.insert(..)` part should occur inside the join, but the `PossibleBorrowerMap` can't be accessed from there.